### DISCUSSION TITLE
Build for Unity Android

### DISCUSSION THREAD

**Discussion by Rickleung** [asus4/onnxruntime-unity-examples#12] [Q&A]
Hi all,

I am trying to implement the onnxruntime into unity android build. However, I encountered many errors after tried to change the .aar to .zip then unzip it. I moved the folder that contained the .so file into unity. Then i got "A failure occurred while executing com.android.build.gradle.internal.tasks.MergeNativeLibsTask$MergeNativeLibsTaskWorkAction". I need help pls...

**Comment by asus4** [comment]
Hi @Rickleung, In recent Unity versions, the `.aar` can be built without unzipping it. What is your use case for using `.so` files?

**Reply by Rickleung** [reply]
the problem is that unity complained about no .so file was found and it couldn't finish the android build. Thus I followed: Download the [onnxruntime-android](https://mvnrepository.com/artifact/com.microsoft.onnxruntime/onnxruntime-android) ( full package) or [onnxruntime-mobile](https://mvnrepository.com/artifact/com.microsoft.onnxruntime/onnxruntime-mobile) ( mobile package) AAR hosted at MavenCentral, change the file extension from .aar to .zip, and unzip it. Include the header files from the headers folder, and the relevant libonnxruntime.so dynamic library from the jni folder in your NDK project. Then more errors come up.

**Reply by asus4** [reply]
The .aar should be imported when you imported "com.github.asus4.onnxruntime" package via UPM correctly. Do you see any errors when opening the "onnxruntime-unity-examples" project?

Also, please check if the aar file exists at the path `onnxruntime-unity-examples/Library/PackageCache/com.github.asus4.onnxruntime@0.1.12/Plugins/Android/onnxruntime-android.aar`.

**Reply by Rickleung** [reply]
i downloaded the zip version of your project then import to my exist project. onnxruntime-android.aar is inside the onnxruntime-unity-examples/Library/PackageCache/com.github.asus4.onnxruntime@0.1.12/Plugins/Android/ folder. I am using 2022.3.19f1. Tried 20f1 as well but no luck. Regarding use the git link, I tried to add package from link: https://github.com/asus4/onnxruntime-unity-examples.git but it said error adding package.

**Reply by Rickleung** [reply]
D:\Unity_Project\Jim_AR_2\Library\PackageCache\com.github.asus4.onnxruntime@0.1.12\Plugins\Android\onnxruntime-android.aar

**Reply by Rickleung** [reply]
I can built the apk if i remove the .so files. However, i am getting this from my android phone: AndroidPlayer "asus_ASUS_I001DA" Failed to load native plugin: Unable to lookup library path for 'onnxruntime.so'. I am using android API 28.

**Reply by Rickleung** [reply]
So unity can found the .so files indeed but I dun know why my android phone was unable to read it, * What went wrong:
Execution failed for task ':launcher:mergeDebugNativeLibs'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.MergeNativeLibsTask$MergeNativeLibsTaskWorkAction
   > 2 files found with path 'lib/arm64-v8a/libonnxruntime.so' from inputs:
      - D:\Unity_Project\Jim_AR_2\Library\Bee\Android\Prj\IL2CPP\Gradle\unityLibrary\build\intermediates\library_jni\debug\jni\arm64-v8a\libonnxruntime.so
      - C:\Users\fcleu\.gradle\caches\transforms-3\96469077c2c7bb751c71cfd9bc40b714\transformed\onnxruntime-android\jni\arm64-v8a\libonnxruntime.so
     If you are using jniLibs and CMake IMPORTED targets, see
     https://developer.android.com/r/tools/jniLibs-vs-imported-targets
